### PR TITLE
osd: add asock command to dump the scrub queue

### PR DESCRIPTION
**Pull request by Liuchang0812**
* add a public method to dumps scrubs
* register new command to asock_admin

any comments is appreciated. 

@tchaikov @liewegas @athanatos 

Fixes: http://tracker.ceph.com/issues/17861

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

### CODE REVIEW

**Comment by Liuchang0812** [comment]
FYI, result of this command:

```
[{"pgid":"1.0","sched_time":"2017-01-02 01:35:23.289862","deadline":"2017-01-07 23:41:25.673073"},{"pgid":"2.6","sched_time":"2017-01-02 04:07:40.349184","deadline":"2017-01-08 03:24:07.231452"},{"pgid":"1.3","sched_time":"2017-01-02 06:18:40.805239","deadline":"2017-01-08 00:14:06.916350"},{"pgid":"2.4","sched_time":"2017-01-02 07:58:41.714257","deadline":"2017-01-08 07:15:31.032673"},{"pgid":"2.3","sched_time":"2017-01-02 09:17:15.723569","deadline":"2017-01-08 05:44:35.507823"},{"pgid":"1.7","sched_time":"2017-01-02 10:02:50.532476","deadline":"2017-01-08 07:58:34.595284"},{"pgid":"1.4","sched_time":"2017-01-02 10:40:38.610703","deadline":"2017-01-08 08:41:33.200372"},{"pgid":"2.7","sched_time":"2017-01-02 11:56:48.821839","deadline":"2017-01-08 04:00:31.616621"},{"pgid":"1.6","sched_time":"2017-01-02 14:09:30.878554","deadline":"2017-01-08 03:37:02.696721"},{"pgid":"2.5","sched_time":"2017-01-02 14:29:59.059526","deadline":"2017-01-08 03:59:52.597162"},{"pgid":"0.7","sched_time":"2017-01-02 18:16:37.088916","deadline":"2017-01-08 08:09:10.976245"}]
```

File: src/osd/OSD.h [hunk 1]
```diff
@@ -685,6 +685,21 @@ class OSDService {
     return true;
   }
 
+  void dumps_scrub(Formatter *f) {
+    assert(f != nullptr);
+    Mutex::Locker l(sched_scrub_lock);
+
+    f->open_array_section("sched scrubs");
```
Comment (tchaikov) [inline src/osd/OSD.h]: an XML tag cannot contain space in it. see https://www.w3.org/TR/2008/REC-xml-20081126/#NT-NameChar.

so probably we can simply put "scrubs" as the tag name?

File: src/osd/OSD.h [hunk 2]
```diff
@@ -686,7 +686,7 @@ class OSDService {
   }
 
   void dumps_scrub(Formatter *f) {
-    assert(f!=nullptr);
+    assert(f != nullptr);
```
Comment (tchaikov) [inline src/osd/OSD.h]: please squash this commit into the previous one.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -2395,8 +2397,16 @@ void OSD::final_init()
...
+  r = admin_socket->register_command("dump_objectstore_kv_stats",
+				     "dump_objectstore_kv_stats",
+				     asok_hook,
+				     "print statistics of kvdb which used by bluestore");
+  assert(r == 0);
+
+  r = admin_socket->register_command("dump_scrubs",
+				     "dump_scrubs",
+				     asok_hook,
+				     "print schedlued scrubs");
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: s/schedlued/scheduled/

File: src/osd/OSD.h [hunk 4]
```diff
@@ -685,6 +685,21 @@ class OSDService {
...
+  void dumps_scrub(Formatter *f) {
+    assert(f != nullptr);
+    Mutex::Locker l(sched_scrub_lock);
+
+    f->open_array_section("sched scrubs");
+    for (const auto &i: sched_scrub_pg) {
+      f->open_object_section("scrub");
+      f->dump_stream("pgid") << i.pgid;
+      f->dump_stream("sched_time") << i.sched_time;
+      f->dump_stream("deadline") << i.deadline;
```
Comment (tchaikov) [inline src/osd/OSD.h]: maybe we can have a boolean field named "forced"? see http://tracker.ceph.com/issues/17861#note-5. i know, technically, it's redundant. but it helps user to understand this scrub. @athanatos what do you think?

File: src/osd/OSD.h (hunk 2 above)
Comment (Liuchang0812) [inline src/osd/OSD.h]: ok, i will squash commits after review

File: src/osd/OSD.h (hunk 1 above)
Comment (Liuchang0812) [inline src/osd/OSD.h]: :+1:  ok

File: src/osd/OSD.h (hunk 4 above)
Comment (Liuchang0812) [inline src/osd/OSD.h]: it has been added. by the way, it seems that there isn't not a way to do a force scrub via ceph tools.

**Comment by athanatos** [comment]
This seems reasonable to me.  I'm ok with it if @tchaikov is.

File: src/osd/OSD.h (hunk 4 above)
Comment (tchaikov) [inline src/osd/OSD.h]: both "ceph osd scrub ${osd_num}" and "ceph pg scrub ${pg_num}" force a scrub job.

**Comment by tchaikov** [comment]
@Liuchang0812 

nit, the change happens in "osd" but not in any of the cli tools, might want to change the prefix of the title to "osd:".

**Comment by Liuchang0812** [comment]
@tchaikov OK, i will change title and test again.

**Comment by Liuchang0812** [comment]
:smile: @tchaikov  I get it now

**Comment by Liuchang0812** [comment]
@tchaikov Updated, Waiting for test result.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-01-04_20:12:04-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2017-01-05_07:47:19-rados-wip-kefu-testing---basic-mira/

**Pull request merged**